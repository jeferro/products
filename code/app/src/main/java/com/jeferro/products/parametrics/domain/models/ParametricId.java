package com.jeferro.products.parametrics.domain.models;

import com.jeferro.shared.ddd.domain.models.aggregates.StringIdentifier;

public class ParametricId extends StringIdentifier {

    public ParametricId(String value) {
        super(value);
    }

    public static ParametricId createOf(String value) {
        return new ParametricId(value);
    }
}
