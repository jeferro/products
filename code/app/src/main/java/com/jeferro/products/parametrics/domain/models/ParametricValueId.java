package com.jeferro.products.parametrics.domain.models;

import com.jeferro.shared.ddd.domain.models.aggregates.StringIdentifier;

public class ParametricValueId extends StringIdentifier {

    public ParametricValueId(String value) {
        super(value);
    }

    public static ParametricValueId createOf(String value) {
        return new ParametricValueId(value);
    }
}
