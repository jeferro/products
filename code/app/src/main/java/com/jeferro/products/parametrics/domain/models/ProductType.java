package com.jeferro.products.parametrics.domain.models;

import com.jeferro.shared.ddd.domain.models.value_objects.ValueObject;
import com.jeferro.shared.ddd.domain.utils.ValueValidationUtils;
import lombok.Getter;

@Getter
public class ProductType extends ValueObject {

  private ProductTypeId id;

  private String name;

  public ProductType(
	  ProductTypeId id,
	  String name) {
	setId(id);
	setName(name);
  }

  public static ProductType createOf(ProductTypeId id, String name) {
	return new ProductType(id, name);
  }

  public boolean hasId(ProductTypeId id) {
	return this.id.equals(id);
  }

  public void setId(ProductTypeId id) {
	ValueValidationUtils.isNotNull(id, "Id");
	this.id = id;
  }

  public void setName(String name) {
	ValueValidationUtils.isNotBlank(name, "Name");
	this.name = name;
  }
}
