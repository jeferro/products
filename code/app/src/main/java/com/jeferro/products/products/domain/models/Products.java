package com.jeferro.products.products.domain.models;

import com.jeferro.products.shared.domain.models.aggregates.EntityCollection;

import java.util.Arrays;
import java.util.List;

public class Products extends EntityCollection<ProductId, Product> {

    public Products(List<Product> entities) {
        super(entities);
    }

    public Products(List<Product> entities, Integer pageSize, Integer pageNumber, Long totalPages) {
        super(entities, pageSize, pageNumber, totalPages);
    }

    public static Products createOf(Product... products) {
        var entities = Arrays.asList(products);
        return new Products(entities);
    }

    public static Products createOfCriteria(List<Product> entities, ProductCriteria criteria, Long totalEntities) {
        return new Products(entities, criteria.getPageSize(), criteria.getPageNumber(), totalEntities);
    }
}
