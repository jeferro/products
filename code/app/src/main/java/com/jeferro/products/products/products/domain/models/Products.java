package com.jeferro.products.products.products.domain.models;

import com.jeferro.shared.ddd.domain.models.aggregates.EntityCollection;

import java.util.Arrays;
import java.util.List;

public class Products extends EntityCollection<ProductCode, Product> {

    public Products(List<Product> entities) {
        super(entities);
    }

    public Products(List<Product> entities, Integer pageSize, Integer pageNumber, Long totalPages) {
        super(entities, pageSize, pageNumber, totalPages);
    }

    public static Products createOf(Product... products) {
        var entities = Arrays.asList(products);
        return new Products(entities);
    }

    public static Products createOfFilter(List<Product> entities, ProductFilter filter, Long totalEntities) {
        return new Products(entities,
            filter.getPageSize(),
            filter.getPageNumber(),
            totalEntities);
    }
}
