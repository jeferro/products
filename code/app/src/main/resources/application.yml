logging:
  level:
    root: WARN
    com.jeferro: INFO

spring:
  application:
    name: products

  profiles:
    active: local
    include:
      - shared
      - kafka
      - mongo

  threads:
    virtual:
      enabled: true
