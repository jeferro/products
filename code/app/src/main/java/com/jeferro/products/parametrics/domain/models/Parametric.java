package com.jeferro.products.parametrics.domain.models;

import com.jeferro.products.parametrics.domain.exceptions.ParametricValueNotFoundException;
import com.jeferro.products.parametrics.domain.models.values.ParametricValueId;
import com.jeferro.products.parametrics.domain.models.values.ParametricValues;
import com.jeferro.shared.ddd.domain.models.projection.Projection;
import com.jeferro.shared.ddd.domain.utils.ValueValidationUtils;
import com.jeferro.shared.locale.domain.models.LocalizedField;
import lombok.Getter;

@Getter
public class Parametric extends Projection<ParametricId> {

  private LocalizedField name;

  private ParametricValues values;

  public Parametric(
	  ParametricId id,
	  LocalizedField name,
	  ParametricValues values) {
	super(id);

	setName(name);
	setValues(values);
  }

  public static Parametric createOf(ParametricId id, LocalizedField name, ParametricValues values) {
	return new Parametric(id, name, values);
  }

  public boolean validate(ParametricValueId parametricValueId) {
	if (notContainsParametricValue(parametricValueId)) {
	  throw ParametricValueNotFoundException.createOf(this, parametricValueId);
	}

	return true;
  }

  private boolean notContainsParametricValue(ParametricValueId parametricValueId) {
	return values.stream()
		.noneMatch(parametricValue -> parametricValue.hasSameId(parametricValueId));
  }

  public void setName(LocalizedField name) {
	ValueValidationUtils.isNotNull(name, "name", this);
	this.name = name;
  }

  private void setValues(ParametricValues values) {
	ValueValidationUtils.isNotEmpty(values, "values", this);
	this.values = values;
  }
}
