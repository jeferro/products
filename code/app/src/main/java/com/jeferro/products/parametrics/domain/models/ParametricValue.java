package com.jeferro.products.parametrics.domain.models;

import java.util.List;

import com.jeferro.shared.ddd.domain.models.projection.Projection;
import com.jeferro.shared.ddd.domain.utils.ValueValidationUtils;
import com.jeferro.shared.locale.domain.models.LocalizedField;
import lombok.Getter;

@Getter
public class ParametricValue extends Projection<ParametricValueId> {

  private LocalizedField name;

  private String value;

  private List<ParametricValue> values;

  public ParametricValue(
	  ParametricValueId id,
	  LocalizedField name,
	  String value,
	  List<ParametricValue> values) {
	super(id);

	setName(name);
	setValue(value);
	setValues(values);
  }

  public static ParametricValue createSimple(ParametricValueId id, LocalizedField name) {
	return new ParametricValue(id, name, null, null);
  }

  private void setName(LocalizedField name) {
	ValueValidationUtils.isNotNull(name, "name", this);
	this.name = name;
  }

  private void setValue(String value) {
	this.value = value;
  }

  private void setValues(List<ParametricValue> values) {
	this.values = values;
  }
}
