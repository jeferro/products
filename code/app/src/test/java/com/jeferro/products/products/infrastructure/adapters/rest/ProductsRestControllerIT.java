package com.jeferro.products.products.infrastructure.adapters.rest;

import com.jeferro.products.products.domain.models.ProductMother;
import com.jeferro.products.products.domain.models.Products;
import com.jeferro.products.shared.application.StubHandlerBus;
import com.jeferro.products.shared.infrastructure.adapters.rest.RestControllerIT;
import com.jeferro.products.shared.infrastructure.adapters.utils.ApprovalUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

@WebMvcTest(ProductsRestController.class)
class ProductsRestControllerIT extends RestControllerIT {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private StubHandlerBus stubHandlerBus;

    @Test
    void givenProducts_whenListProducts_thenExecutesListProductsCommand() throws Exception {
        var products = Products.createOf(
                ProductMother.apple(),
                ProductMother.pear()
        );
        stubHandlerBus.init(products);

        var requestBuilder = MockMvcRequestBuilders.get("/v1/products")
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_CONTENT);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstCommandOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void givenAnProduct_whenCreateProduct_thenExecutesCreateProductCommand() throws Exception {
        var apple = ProductMother.apple();
        stubHandlerBus.init(apple);

        var requestContent = """
                {
                  "name": "%s"
                }"""
                .formatted(apple.getName());

        var requestBuilder = MockMvcRequestBuilders.post("/v1/products")
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_CONTENT)
                .content(requestContent);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstCommandOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void givenAnProduct_whenGetProduct_thenExecutesGetProductCommand() throws Exception {
        var apple = ProductMother.apple();
        stubHandlerBus.init(apple);

        var requestBuilder = MockMvcRequestBuilders.get("/v1/products/" + apple.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_CONTENT);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstCommandOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void givenAnProduct_whenUpdateProduct_thenExecutesUpdateProductCommand() throws Exception {
        var apple = ProductMother.apple();
        stubHandlerBus.init(apple);

        var requestContent = """
                {
                  "name": "%s"
                }"""
                .formatted(apple.getName());

        var requestBuilder = MockMvcRequestBuilders.put("/v1/products/" + apple.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_CONTENT)
                .content(requestContent);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstCommandOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

    @Test
    void givenAnProduct_whenDeleteProduct_thenExecutesDeleteProductCommand() throws Exception {
        var apple = ProductMother.apple();
        stubHandlerBus.init(apple);

        var requestBuilder = MockMvcRequestBuilders.delete("/v1/products/" + apple.getId())
                .contentType(MediaType.APPLICATION_JSON)
                .header(HttpHeaders.AUTHORIZATION, AUTHORIZATION_CONTENT);

        var response = mockMvc.perform(requestBuilder)
                .andReturn()
                .getResponse();

        ApprovalUtils.verifyAll(stubHandlerBus.getFirstCommandOrError(),
                response.getStatus(),
                response.getContentAsString());
    }

}