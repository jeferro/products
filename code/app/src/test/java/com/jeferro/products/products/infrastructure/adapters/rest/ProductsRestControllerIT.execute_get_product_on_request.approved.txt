data[0] = GetProductCommand[productId=apple,auth=user]
data[1] = 200
data[2] = {"id":"apple","name":"Apple"}
